doc/DataPlaneCodeGeneration/Handle-Ambiguity.md

```diff
@@ -0,0 +1,225 @@
+```
+What we try to generate:
+```C#
+public virtual async Task<Response<Model>> OperationAsync(string requiredId, string optionalId = null, CancellationToken cancellationToken = default) // Convenience method
+{
+    ...
+}
+
+public virtual async Task<Response> OperationAsync(string requiredId, string optionalId = null, RequestContext context = null) // Protocol method
+{
+    ...
+}
+```
+It only has one required parameter `requiredId`, so the call that we guarantee will not have ambiguity is
+```C#
+OperationAsync("requiredId");
+```
+Then we find that this call has ambiguity error, so we have to change our signature by choosing from [one of these solutions](#solution).
+
+### Operation with input
+Spec:
+```
+model Model {}
+op operation(@body model: Model): Model;
+```
+```C#
+public virtual async Task<Response<Model>> OperationAsync(Model model, CancellationToken cancellationToken = default) // Convenience method
+{
+    ...
+}
+
+public virtual async Task<Response> OperationAsync(RequestContent content, RequestContext context = null) // Protocol method
+{
+    ...
+}
+```
+No matter if the input model is required or optional in spec, it is a required parameter in the signature. So, what we care about are:
+```C#
+OperationAsync(new Model());
+OperationAsync(new RequestContent());
+```
+Then we find out these calls are fine, so just keep these signatures.
+
+### Operation without both input and output
+Spec:
+```
+op operation(@path requiredId: string, @path optionalId?: string): void;
+```
+What we try to generate:
+```C#
+public virtual async Task<Response> OperationAsync(string requiredId, string optionalId = null, RequestContext context = null) // Protocol method
+{
+    ...
+}
+```
+We don't generate convenience method for this case (see [this](#convenience-method-not-always-generated) to see why), so there will be no ambiguity.
+### Operation with spread scenario
+Spec:
+```
+alias OperationAlias = {
+  requiredId: string;
+  optionalId?: string;
+};
+op operation(...OperationAlias): void;
+```
+What we try to generate:
+```C#
+public virtual async Task<Response> OperationAsync(string requiredId, string optionalId = null, CancellationToken cancellationToken = default) // Convenience method
+{
+    ...
+}
+public virtual async Task<Response> OperationAsync(RequestContent content, RequestContext context = null) // Protocol method
+{
+    ...
+}
+```
+The convenience method has a required parameter `requiredId` and protocol method has a required method `content`, so what we guarantee will not have ambiguity are
+```C#
+OperationAsync("requiredId");
+OperationAsync(new RequestContent());
+```
+Then we find these calls are fine, so just keep these signatures.
+
+## Appendix
+### Convenience method not always generated
+There is one [situation](#operation-without-both-input-and-output) we don't generate convenience method. Think about what would be like if we generate the convenience method.
```

Comment:
```suggestion
There is one [situation](#operation-without-both-input-and-output) we don't generate convenience method, because if we did, the convenience method would have almost exact the same parameter list as the protocol method:
```